readme/coding_style.md

```diff
@@ -0,0 +1,247 @@
+export default const processData = () => {
+	// ...
+};
+
+// foo.ts
+import processData from './processData';
+
+processData();
+...
+```
+
+## Use `camelCase` for `const`ants in new code
+
+**BAD:**
+```ts
+// Bad! Don't use in new code!
+const GRAVITY_ACCEL = 9.8;
+```
+
+**Good:**
+```ts
+const gravityAccel = 9.8;
+```
+
+
+## Indent using `tab`s
+
+**VSCode**: In `vscode`, be sure to check whether new files are created with `tab` or `space` indentation! [Spaces can be converted to tabs using the command palette.](https://code.visualstudio.com/docs/editor/codebasics#_autodetection)
+
+
+## Use strict equality
+
+Use `===` instead of `==`.
+
+Although the TypeScript compiler _will_ give error messages if two different types are compared with `==` (e.g. `number == boolean`), its compiler error [messages in this case can be misleading](https://github.com/microsoft/TypeScript/issues/26592).
+
+
+### See also
+ * [Unofficial TypeScript style guide, `==` vs `===`](https://basarat.gitbook.io/typescript/styleguide#or)
+ * [More about `==` vs `===` in TypeScript.](https://stackoverflow.com/a/60669874)
+
+
+## Declare variables just before their usage
+
+**BAD:**
+```ts
+// Bad!
+let foo, bar;
+
+const doThings = () => {
+	// do things unrelated to foo, bar
+};
+
+// Do things involving foo and bar
+foo = Math.random();
+bar = foo + Math.random() / 100;
+foo += Math.sin(bar + Math.tan(foo));
+...
+```
+
+**Good:**
+```ts
+...
+const doThings = () => {
+	// do things unrelated to foo, bar
+};
+
+// Do things involving foo and bar
+let foo = Math.random();
+let bar = foo + Math.random() / 100;
+foo += Math.sin(bar + Math.tan(foo));
+...
+```
+
+Don't allow this to lead to duplicate code, however. If constants are used multiple times, it's okay to declare them at the top of a file or in a separate, imported file.
+
+
+## Prefer `const` to `let` (where possible)
+
+
+## Prefer `() => {}` to `function() { ... }`
+
+Doing this avoids having to deal with the `this` keyword. Not having it makes it easier to refactor class components into React Hooks, because any use of `this` (used in classes) will be correctly detected as invalid by TypeScript.
+
+**BAD:**
+```ts
+// Bad!
+function foo() {
+	...
+}
+```
+
+**Good:**
+```ts
+const foo = () => {
+	...
+};
+```
+
+### See also
+ * [Frontend Armory — When should I use arrow functions with React?](https://frontarm.com/james-k-nelson/when-to-use-arrow-functions/)
+
+
+
+## Avoid default and optional parameters
+
+As much as possible, avoid default parameters in **function definitions** and optional fields in **interface definitions**. When all parameters are required, it is much easier to refactor the code because the compiler will automatically catch any missing parameters.
+
+**BAD:**
+```ts
+interface CustomTag {
+	name: string;
+	validContentValues?: string[];
+}
+
+const createTagFromName = (name: string, validContentValues: string[] = []): CustomTag => {
+	return {
+		name,
+		validContentValues,
+	};
+};
```

Comment:
<!--In Rust, a pattern similar to the following [is common](https://doc.rust-lang.org/rust-by-example/fn/methods.html):

<details><summary>Rust version</summary>

```rust
struct Color {
  r: f64,
  g: f64,
  b: f64,
}

impl Color {
  fromRGB(r: f64, g: f64, b: f64) -> Color {
    ...
  }

  fromHSV(hue: f64, saturation: f64, luminance: f64) -> Color {
    ...
  }
}
```

</details>

```ts
interface Color {
  red: number;
  green: number;
  blue: number;
}

namespace Color {
  export const fromRGB = (red: number, green: number, blue: number): Color => {
    ...
  };

  ...
}
```

While this works, I don't like it in JS. How should constructors for interfaces be defined? Are classes always preferable to interfaces in cases like this?-->

This code should probably be a `class` with static constructor methods and not an `interface`.